src/devices/MatrixKeyboard/MatrixKeyboard.cs

```diff
@@ -0,0 +1,231 @@
+                    }
+                }
+                catch (Exception ex)
+                {
+                    IsScanning = false;
+                    throw new Exception("Error while openning GPIO pins.", ex);
+                }
+
+                IsScanning = true;
+                try
+                {
+                    _scanThread.Start();
+                }
+                catch (Exception ex)
+                {
+                    IsScanning = false;
+                    throw new Exception("Error while starting scanning thread.", ex);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Stop keyboard scanning then close GPIO pins
+        /// </summary>
+        public void StopScan()
+        {
+            if (IsScanning)
+            {
+                IsScanning = false;
+                Thread.Sleep(_scanInterval); // wait for thread ends
+
+                try
+                {
+                    for (var i = 0; i < _rowPins.Length; i++)
+                    {
+                        _gpio.ClosePin(_rowPins[i]);
+                    }
+                    for (var i = 0; i < _columnPins.Length; i++)
+                    {
+                        _gpio.ClosePin(_columnPins[i]);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    IsScanning = false;
+                    throw new Exception("Error while closing GPIO pins.", ex);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get all buttons' values
+        /// </summary>
+        public ReadOnlySpan<PinValue> Values => _buttonValues.AsSpan();
+
+        /// <summary>
+        /// Get buttons' values by row
+        /// </summary>
+        /// <param name="row">Row index</param>
+        public ReadOnlySpan<PinValue> RowValues(int row) => _buttonValues.AsSpan(row * _rowPins.Length, _columnPins.Length);
+
+        /// <summary>
+        /// Keyboard event
+        /// </summary>
+        public event PinChangeEventHandler PinChangeEvent;
+
+        /// <summary>
+        /// Delegate of keyboard event
+        /// </summary>
+        public delegate void PinChangeEventHandler(object sender, MatrixKeyboardEventArgs pinValueChangedEventArgs);
+
+        public void Dispose()
+        {
+            IsScanning = false;
+            _scanThread = null;
+
+            _gpio?.Dispose();
+            _gpio = null;
+        }
+
+        private IGpioController _gpio;
+        private Thread _scanThread;
+        private TimeSpan _scanInterval;
+        private TimeSpan CalcScanInterval(double scanFrequency)
+        {
+            // "scanFrequency * _rowPins.Length" is how many rows to scan per second
+            // it divide by 10000000(ticks) so we can get interval of the timer
+            return new TimeSpan((long)(10000000 / (scanFrequency * _rowPins.Length)));
+        }
+        private void ScanThreadStart()
```

Comment:
would it make sense to have:

```
Task ScanAsync(CancellationToken); // for starting and managing the loop
Task ReadKeyAsync(); // loop would fill up the queue of keys so this could either return immediately or wait for the loop to add key to the queue
```